File: Reorder Data in Log Files.py
""""// Time Complexity : O(nk log(k))
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
"""


class Solution:
    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        def customSort(a, b):
            strArr1 = a.split(' ', 1)
            strArr2 = b.split(' ', 1)
            if not strArr1[1][0].isdigit() and not strArr2[1][0].isdigit():
                if strArr1[1] < strArr2[1]:
                    return -1
                elif strArr1[1] == strArr2[1]:
                    if strArr1[0] < strArr2[0]:
                        return -1
                    else:
                        return 1
                else:
                    return 1

            elif strArr1[1][0].isdigit() and not strArr2[1][0].isdigit():
                return 1
            elif not strArr1[1][0].isdigit() and strArr2[1][0].isdigit():
                return -1
            else:
                return 0

        logs.sort(key=cmp_to_key(customSort))
        return logs

# class Solution:
#     def reorderLogFiles(self, logs: List[str]) -> List[str]:
#         letters = []
#         numbers=[]
#         for i in logs:
#             if i.split()[1].isdigit():
#                 numbers.append(i)
#             else:
#                 letters.append(i)


#         letters.sort(key = lambda x:x.split()[0])
#         letters.sort(key = lambda x:x.split()[1:])

#         result = letters + numbers

#         return result

File: problem3/ReorderDataInLogFiles.java
// Time Complexity : O(knlog n), n -> Number of logs, k -> Average length of each log
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem3;

import java.util.Arrays;

public class ReorderDataInLogFiles {
	public String[] reorderLogFiles(String[] logs) {
		if (logs == null || logs.length < 2) {
			return logs;
		}

		Arrays.sort(logs, (l1, l2) -> {
			String[] strArr1 = l1.split(" ", 2);
			String[] strArr2 = l2.split(" ", 2);

			boolean isDigitLog1 = Character.isDigit(strArr1[1].charAt(0));
			boolean isDigitLog2 = Character.isDigit(strArr2[1].charAt(0));

			if (!isDigitLog1 && !isDigitLog2) {
				int comp = strArr1[1].compareTo(strArr2[1]);
				if (comp == 0) {
					return strArr1[0].compareTo(strArr2[0]);
				}
				return comp;
			} else if (isDigitLog1 && !isDigitLog2) {
				return 1;
			} else if (!isDigitLog1 && isDigitLog2) {
				return -1;
			} else {
				return 0;
			}
		});

		return logs;
	}

	private void print(String[] logs) {
		StringBuilder sb = new StringBuilder();
		for (String log : logs) {
			if (sb.length() != 0) {
				sb.append(", ");
			}
			sb.append(log);
		}

		System.out.print(sb);
	}

	public static void main(String[] args) {
		ReorderDataInLogFiles obj = new ReorderDataInLogFiles();
		String[] logs = { "dig1 8 1 5 1", "let1 art can", "dig2 3 6", "let2 own kit dig", "let3 art zero" };
		System.out.print("Logs: ");
		obj.print(logs);
		System.out.println();

		String[] reorderedLogs = obj.reorderLogFiles(logs);
		System.out.print("Reordered Logs: ");
		obj.print(reorderedLogs);
	}

}
